File: ProfilesReviewEmailLibrary/ReviewEmail.vb
Imports System.Configuration
Imports System.Net.Mail
Imports System.Text
Imports ProfilesReviewEmailLibrary

Public Module ReviewEmail

    Public Sub SendReviewEmails(ByVal emailSubjectLine As String, _
                                ByVal emailFromAddress As String, _
                            ByVal emailFromDisplayName As String, _
                            ByVal smtpServer As String, _
                            ByVal smtpPort As String, _
                            ByVal externalWebsiteURL As String, _
                            ByVal internalWebsiteURL As String)

        Dim emailUserInfoList As EmailUserInfoList
        emailUserInfoList = emailUserInfoList.GetEmailInfoList()

        If emailUserInfoList IsNot Nothing And emailUserInfoList.Count > 0 Then
            For Each emailUserInfo As EmailUserInfo In emailUserInfoList
                SendMessage(emailSubjectLine, _
                            emailUserInfo, _
                            emailFromAddress, _
                            emailFromDisplayName, _
                            smtpServer, _
                            smtpPort, _
                            externalWebsiteURL, _
                            internalWebsiteURL)
            Next
        End If

    End Sub

    Private Sub SendMessage(ByVal emailSubjectLine As String, _
                            ByVal emailUserInfo As EmailUserInfo, _
                            ByVal emailFromAddress As String, _
                            ByVal emailFromDisplayName As String, _
                            ByVal smtpServer As String, _
                            ByVal smtpPort As String, _
                            ByVal externalWebsiteURL As String, _
                            ByVal internalWebsiteURL As String)

        Dim msg As New MailMessage()

        msg.Sender = New MailAddress(emailFromAddress, emailFromDisplayName)
        msg.From = New MailAddress(emailFromAddress, emailFromDisplayName)
        msg.To.Add(New MailAddress(emailUserInfo.Email, emailUserInfo.UserName))
        msg.Priority = MailPriority.Normal
        msg.Subject = emailSubjectLine
        msg.Body = BuildMessageBody(emailUserInfo.FullName, _
                                    emailUserInfo.ProfileInfoList, _
                                    externalWebsiteURL, _
                                    internalWebsiteURL)
        msg.IsBodyHtml = True

        SendMail(msg, smtpServer, smtpPort)

    End Sub

    Private Function SendMail(ByVal msg As MailMessage, _
                              ByVal smptServerName As String, _
                              ByVal smptPort As String) As Boolean

        Try
            Dim smtpClient As New SmtpClient(smptServerName, smptPort)
            smtpClient.Send(msg)
            Return True
        Catch ex As Exception
            'TODO: write error to log.
            'Diagnostics.EventLog.WriteEntry(My.Settings.AppName, ex.Message.ToString, EventLogEntryType.Error)
            Return False
        End Try

        Return True
    End Function

    Private Function BuildMessageBody(ByVal name As String, _
                                      ByVal profileInfoList As ProfileInfoList, _
                                      ByVal externalWebsiteURL As String, _
                                      ByVal internalWebsiteURL As String) As String

        Dim emailBodyBuilder As New StringBuilder

        emailBodyBuilder.Append("Dear " & name & ", ")
        emailBodyBuilder.Append("<br/><br/>")
        emailBodyBuilder.Append("Please find the list of approaching, overdue and completed author, VA and policy reviews below. ")
        emailBodyBuilder.Append("After completing an author, VA or policy review please update the relevant review status for the profile section in the system. ")
        emailBodyBuilder.Append("This can be achieved by locating the profile on the search page, clicking the browse link and then selecting the revision dates tab. ")
        emailBodyBuilder.Append("Set the status to complete by using the drop down list. ")
        emailBodyBuilder.Append("<br/><br/>")

        If profileInfoList IsNot Nothing And profileInfoList.Count > 0 Then

            For Each profile As ProfileInfo In profileInfoList

                If String.IsNullOrEmpty(profile.ScenarioTitle) Then
                    emailBodyBuilder.Append("<strong><u>" & profile.ProfileTitle & " - v" & profile.FullVersion & " (Draft)" & "</u></strong>")
                Else
                    emailBodyBuilder.Append("<strong><u>" & profile.ProfileTitle & " (" & profile.ScenarioTitle & ") " & "- v" & profile.FullVersion & " (Draft)" & "</u></strong>")
                End If

                emailBodyBuilder.Append("<br/><br/>")
                emailBodyBuilder.Append("<table border=1 cellpadding=0 cellspacing=0>")

                emailBodyBuilder.Append("<tr>")
                emailBodyBuilder.Append(BuildCell(String.Empty))
                emailBodyBuilder.Append(BuildCell("Section name", True, "#000000", "verdana", 1, "center"))
                emailBodyBuilder.Append(BuildCell("VA review", True, "#000000", "verdana", 2, "center"))
                emailBodyBuilder.Append(BuildCell("Policy review", True, "#000000", "verdana", 2, "center"))
                emailBodyBuilder.Append(BuildCell("Author review", True, "#000000", "verdana", 2, "center"))
                emailBodyBuilder.Append("</tr>")

                Dim currentDate As New DateTime
                currentDate = DateTime.Now()

                For Each section As SectionInfo In profile.SectionInfoList
                   
                    emailBodyBuilder.Append("<tr>")
                    emailBodyBuilder.Append(BuildCell(section.SectionNumber.ToString()))
                    emailBodyBuilder.Append(BuildCell(section.SectionName))

                    BuildDateCells(section.VAReviewDate, section.VAReviewStatus, currentDate, emailBodyBuilder)

                    BuildDateCells(section.PolicyReviewDate, section.PolicyReviewStatus, currentDate, emailBodyBuilder)

                    BuildDateCells(section.AuthorReviewDate, section.AuthorReviewStatus, currentDate, emailBodyBuilder)

                    emailBodyBuilder.Append("</tr>")

                Next

                emailBodyBuilder.Append("</table>")
                emailBodyBuilder.Append("<br />")

            Next

        End If

        emailBodyBuilder.Append("<br/><br/>")
        emailBodyBuilder.Append("The D2R2 system can be accessed via the internet using the following link: ")
        emailBodyBuilder.Append(externalWebsiteURL)
        emailBodyBuilder.Append("<br/><br/>")
        emailBodyBuilder.Append("Internal users can access the D2R2 system using the following link: ")
        emailBodyBuilder.Append(internalWebsiteURL)
        emailBodyBuilder.Append("<br/><br/>")
        emailBodyBuilder.Append("If you have a problem accessing the D2R2 application (or aren't sure why you've received this email) please contact us by emailing the D2R2 mail box: D2R2@defra.gsi.gov.uk")
        emailBodyBuilder.Append("<br/><br/>")
        Return emailBodyBuilder.ToString()

    End Function

    Private Sub BuildDateCells(ByVal dateTime As DateTime, ByVal reviewStatus As String, ByVal currentDate As DateTime, ByVal emailBodyBuilder As StringBuilder)

        Dim twoWeeksDate As New DateTime
        twoWeeksDate = currentDate.AddDays(14)

        If dateTime < currentDate Or dateTime < twoWeeksDate Then
            emailBodyBuilder.Append(BuildCell(dateTime.ToString("dd/MM/yyyy")))
            If reviewStatus.Equals("Complete") Then
                emailBodyBuilder.Append(BuildCell(" (Completed) ", True, "#009900"))
            Else
                If dateTime < currentDate Then
                    emailBodyBuilder.Append(BuildCell(" (Overdue) ", True, "#FF0000"))
                Else
                    emailBodyBuilder.Append(BuildCell(" (Due soon) ", True, "#FF6103"))
                End If
            End If
        Else
            emailBodyBuilder.Append(BuildEmptyCells())
        End If

    End Sub

    Private Function BuildEmptyCells()
        Return BuildCell(String.Empty, False, "#000000", "verdana", 2, "left")
    End Function

    Private Function BuildCell(ByVal text As String)
        Return BuildCell(text, False, "#000000", "verdana", 1, "left")
    End Function

    Private Function BuildCell(ByVal text As String, ByVal isBold As Boolean)
        Return BuildCell(text, isBold, "#000000", "verdana", 1, "left")
    End Function

    Private Function BuildCell(ByVal text As String, ByVal isBold As Boolean, ByVal colour As String)
        Return BuildCell(text, isBold, colour, "verdana", 1, "left")
    End Function

    Private Function BuildCell(ByVal text As String, _
                               ByVal isBold As Boolean, _
                               ByVal colour As String, _
                               ByVal font As String, _
                               ByVal colSpan As Integer, _
                               ByVal align As String) As String

        Dim sb As New StringBuilder
        sb.Append("<td align=" & align & " colspan=" & colSpan.ToString() & ">")

        If isBold Then
            sb.Append("<strong>")
        End If

        sb.Append("<font size=2 face=" & font & " color=" & colour & ">")
        sb.Append(text)
        sb.Append("</font>")

        If isBold Then
            sb.Append("</strong>")
        End If
        sb.Append("</td>")
        Return sb.ToString()

    End Function

End Module

File: Profiles.Business/EmailBusiness/Review/ReviewEmail.cs
using System;
using System.Collections.Generic;
using System.Net.Mail;
using Profiles.EmailService;
using Profiles.Infrastructure.Enums;
using Profiles.Contracts.DataContracts;
using ProfilesLibrary;
using System.IO;

namespace Profiles.Business.EmailBusiness.Review.Email
{
    public class ReviewEmail : IEmail<UserDueReviewEmailResponse>
    {
        public string Subject
        {
            get { return "D2R2 - Profile review request"; }
        }

        public MailPriority Priority
        {
            get { return MailPriority.Normal; }
        }

        public string GeneratePlainTextContent(UserDueReviewEmailResponse data)
        {
            return string.Empty;
        }

        public string GenerateHtmlContent(UserDueReviewEmailResponse data)
        {
            // Header
            var contentMeta = new HtmlTagHelper("meta");
            contentMeta.AddAttribute("content", "utf-8");
            contentMeta.AddAttribute("http-equiv", "encoding");

            var encodingMeta = new HtmlTagHelper("meta");
            encodingMeta.AddAttribute("content", "utf-8");
            encodingMeta.AddAttribute("http-equiv", "encoding");

            var title = new HtmlTagHelper("title");
            title.AppendInnerContent(Subject);

            var header = new HtmlTagHelper("head", contentMeta, encodingMeta, title);

            // Body
            var body = new HtmlTagHelper("body");

            var dear = new HtmlTagHelper("p");
            dear.AppendInnerContent("Dear " + data.FullName + ",");

            body.AppendInnerContent(dear);

            var firstParagraph = new HtmlTagHelper("p");
            firstParagraph.AppendInnerContent("Please find the list of approaching, overdue and completed author, VA and policy reviews below. ");
            firstParagraph.AppendInnerContent(
                "After completing an author, VA or policy review please update the relevant review status for the profile section in the system. ");
            firstParagraph.AppendInnerContent(
                "This can be achieved by locating the profile on the search page, clicking the browse link and then selecting the revision dates tab. ");
            firstParagraph.AppendInnerContent("Set the status to complete by using the drop down list. ");

            body.AppendInnerContent(firstParagraph);

            foreach (var profileVersion in data.ProfileVersions)
            {
                var profileTitle = new HtmlTagHelper("u");

                if (string.IsNullOrEmpty(profileVersion.ScenarioTitle))
                {
                    profileTitle.AppendInnerContent(string.Format("{0} - v {1} (Draft)", profileVersion.ProfileTitle,
                        profileVersion.FullVersion));
                }
                else
                {
                    profileTitle.AppendInnerContent(string.Format("{0} ({1}) - v {2} (Draft)", profileVersion.ProfileTitle,
                        profileVersion.ScenarioTitle, profileVersion.FullVersion));
                }

                body.AppendInnerContent(new HtmlTagHelper("p", new HtmlTagHelper("strong", profileTitle)));
                body.AppendInnerContent(new HtmlTagHelper("br"));

                var table = new HtmlTagHelper("table");
                table.AddAttribute("border", "1");
                table.AddAttribute("cellpadding", "0");
                table.AddAttribute("cellspacing", "0");

                var tableHeaders = new HtmlTagHelper("tr",
                    TableCell(string.Empty),
                    TableCell("Section name", true, 1, "center"),
                    TableCell("VA review", true, 2, "center"),
                    TableCell("Policy review", true, 2, "center"),
                    TableCell("Author review", true, 2, "center"));

                table.AppendInnerContent(tableHeaders);

                foreach (var section in profileVersion.ProfileSections)
                {
                    var sectionNumber = TableCell(section.SectionNumber.ToString());
                    var sectionName = TableCell(section.SectionName);

                    var vaReviewDate = ReviewDateCell(section.NextAuthorReview);
                    var vaReviewStatus = ReviewStatusCell(section.NextAuthorReview, section.AuthorReviewStatus);

                    var technicalReviewDate = ReviewDateCell(section.NextTechnicalReview);
                    var technicalReviewStatus = ReviewStatusCell(section.NextTechnicalReview, section.TechnicalReviewStatus);

                    var policyReviewDate = ReviewDateCell(section.NextPolicyReview);
                    var policyReviewStatus = ReviewStatusCell(section.NextPolicyReview, section.PolicyReviewStatus);

                    table.AppendInnerContent(new HtmlTagHelper("tr",
                        sectionNumber,
                        sectionName,
                        vaReviewDate,
                        vaReviewStatus,
                        technicalReviewDate,
                        technicalReviewStatus,
                        policyReviewDate,
                        policyReviewStatus));
                }

                body.AppendInnerContent(table);
                body.AppendInnerContent(new HtmlTagHelper("br"));
            }

            body.AppendInnerContent(new HtmlTagHelper("br"));
            body.AppendInnerContent(new HtmlTagHelper("br"));
            body.AppendInnerContent(
                "Internal users can access the D2R2 system using the following link: " + GlobalSettings.InternalHomePageUrlForReports);
            body.AppendInnerContent(new HtmlTagHelper("br"));
            body.AppendInnerContent(new HtmlTagHelper("br"));
            body.AppendInnerContent(
                "If you have a problem accessing the D2R2 application (or aren't sure why you've received this email) please contact us by emailing the D2R2 mail box: D2R2@defra.gov.uk");
            body.AppendInnerContent(new HtmlTagHelper("br"));
            body.AppendInnerContent(new HtmlTagHelper("br"));

            return new HtmlTagHelper("html", header, body).ToString();
        }

        public IEnumerable<Attachment> GenerateAttachments(UserDueReviewEmailResponse data)
        {
            foreach (var profileVersion in data.ProfileVersions)
            {
                var profileInfo = ProfileVersionInfo.GetProfileVersionInfo(profileVersion.ProfileVersionId);
                var profile = Profile.GetProfile(profileInfo.ProfileId);

                if (profile.HasCurrentPublishedVersion)
                {
                    var publishedProfileVersionId = profile.CurrentPublishedVersion.Id;

                    if (publishedProfileVersionId != null)
                    {
                        var pdfData = GetProfileVersionComparisonReportCommand.GetProfileVersionComparisonReport(publishedProfileVersionId, profileVersion.ProfileVersionId);
                        var attachmentName = string.Format("{0} {1} - {2} {3} Comparison.pdf", profile.CurrentPublishedVersion.Title, profile.CurrentPublishedVersion.Version, profile.Title, profileInfo.Version);
                        yield return new Attachment(new MemoryStream(pdfData), attachmentName, "application/pdf");
                    }
                }
            }
        }

        private HtmlTagHelper ReviewStatusCell(DateTime? reviewDate, ProfileSectionReviewStatus status)
        {
            if (reviewDate.HasValue && reviewDate.Value.Date < DateTime.Now.AddDays(14))
            {
                switch (status)
                {
                    case ProfileSectionReviewStatus.Complete:
                        return TableCell(" (Completed) ", true, colorHex: "#009900");
                    case ProfileSectionReviewStatus.Active:
                    case ProfileSectionReviewStatus.Proposed:
                        return reviewDate < DateTime.Now.Date
                            ? TableCell(" (Overdue) ", true, colorHex: "#FF0000")
                            : TableCell(" (Due soon) ", true, colorHex: "#FF6103");
                    case ProfileSectionReviewStatus.Unspecified:
                    default:
                        return TableCell();
                }
            }

            return TableCell();
        }

        private HtmlTagHelper ReviewDateCell(DateTime? reviewDate)
        {
            return reviewDate.HasValue && reviewDate.Value.Date < DateTime.Now.AddDays(14)
                ? TableCell(reviewDate.Value.ToString("dd/MM/yyyy"))
                : TableCell();
        }

        private HtmlTagHelper TableCell(string text = "", bool isBold = false, int colSpan = 1, string align ="left", string colorHex = "#000000")
        {
            var cell = new HtmlTagHelper("td");
            cell.AddAttribute("align", align);
            cell.AddAttribute("colspan", colSpan.ToString());

            var content = new HtmlTagHelper("font");
            content.AddAttribute("face", "verdana");
            content.AddAttribute("size", "2");
            content.AddAttribute("color", colorHex);
            content.AppendInnerContent(text);

            cell.AppendInnerContent(isBold
                ? new HtmlTagHelper("strong", content)
                : content);

            return cell;
        }
    }
}
